port/zos390/omrvmem.c

```diff
@@ -1296,24 +1298,31 @@ omrvmem_default_large_page_size_ex(struct OMRPortLibrary *portLibrary, uintptr_t
 	}
 #endif /* OMR_ENV_DATA64 */
 
-	while (0 != PPG_vmem_pageSize[index]) {
-		if (0 != (OMRPORT_VMEM_MEMORY_MODE_EXECUTE & mode)) {
-			if ((ONE_M == PPG_vmem_pageSize[index]) &&
-				(0 != (OMRPORT_VMEM_PAGE_FLAG_PAGEABLE & PPG_vmem_pageFlags[index]))
-			) {
-				break;
-			}
-#if defined(OMR_ENV_DATA64)
-		} else {
-			if ((ONE_M == PPG_vmem_pageSize[index]) &&
-				(0 != (OMRPORT_VMEM_PAGE_FLAG_FIXED & PPG_vmem_pageFlags[index]))
-			) {
-				break;
-			}
-#endif
+	/* Determine maximum pageable, and fixed large pages available. */
+	while (0 != PG_vmem_pageSize[index] && ONE_M >= PPG_vmem_pageSize[index]) {
+
+		if (0 != (OMRPORT_VMEM_PAGE_FLAG_PAGEABLE & PPG_vmem_pageFlags[index]) {
+			/* Pageable. Keep track of largest pageable size */
+			maxPageableIndex = index;
+		}
+		else if (0 != (OMRPORT_VMEM_PAGE_FLAG_FIXED & PPG_vmem_pageFlags[index])) {
+			/* Fixed. Keep track of largest fixed size */
+			maxFixedIndex = index;
 		}
 		index++;
 	}
+
+	/* When mode is OMRPORT_VMEM_MEMORY_EXECUTE we will only consider pagable large pages.
+	 * When flag is OMRPORT_VMEM_PAGE_FLAG_PAGEABLE_PREFERABLE, we prefer pageable unless fixed is larger.
+	 */
+	if (0 != (OMRPORT_VMEM_MEMORY_MODE_EXECUTE & mode) ||
+	   (OMRPORT_VMEM_PAGE_FLAG_PAGEABLE_PREFERABLE == *pageFlags && PPG_vmem_pageSize[maxPageableIndex] >= PPG_vmem_pageSize[maxFixedIndex])) {
+		index = maxPageableIndex;
+	}
+	else {
+		index = maxFixedIndex;
```

Comment:
I am asking again: old code returned 1m fixed for ZOS 64 bit only. I don't see this check in new code. Do I miss something?

Reply:
Thanks for finding that. It was an unintentional oversight. 

I believe OMRPORT_VMEM_MEMORY_MODE_EXECUTE will always be used with 31 bit Z/OS, since we exit out of the API immediately if that's not the case. 

See: https://github.com/eclipse/omr/pull/4762/commits/1e7f2007e1c49c6326a2419ade5a650bf6e97a91